### ISSUE TITLE
A `crossplane.io/pause` annotation to stop reconciliation

### ISSUE THREAD

**Issue by muvaf** [crossplane/crossplane-runtime#351]
### What problem are you facing?

There are cases where users want Crossplane or provider controllers to stop reconciling a certain resource such as migration from a CRD to another (like [here](https://github.com/crossplane-contrib/provider-aws/blob/master/cluster/UPGRADE.md)) or an in-cluster YAML surgery where you don't want to deal with racing conditions. Currently, such operations require creating a `ControllerConfig` for providers to scale down and then shutting down Crossplane by scaling it down. This method stops the world for every resource in the cluster. 

I think this was discussed somewhere but I couldn't find it.

### How could Crossplane help solve your problem?

We can have an controllar-runtime event filter to add to all reconcilers in runtime and core Crossplane to filter out all instances with the annotation `crossplane.io/pause: "true"`. It should be part of the Crossplane Resource Model (XRM) and documented [here](https://github.com/crossplane/crossplane/blob/master/design/one-pager-managed-resource-api-design.md).

**Comment by muvaf** [comment]
@negz FYI, we're planning to tackle this in the next two weeks.

**Comment by negz** [comment]
@muvaf Thanks! Let's make sure to capture this somewhere on the user-facing https://crossplane.io/docs site. In particular I'd like to make sure we're clear that this this isn't really meant to be a replacement for observe only resources (e.g. because a not-reconciled resource is not the same as an observe only resource).

**Comment by negz** [comment]
xref https://github.com/crossplane/crossplane/issues/2590, which is an old duplicate of this issue.

**Comment by ulucinar** [comment]
I've investigated whether we can use controller-runtime's [`predicates`](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/predicate) in the implementation. The idea would be:
- While building the controller for a managed resource, we would install a predicate either as a global event filter, or as an `OwnsOption` (a filter specific to `Owns` watches). This controller-runtime predicate could be defined in crossplane-runtime.
- That predicate then checks for a managed resource, if it's explicitly installed for the managed resource's event handler, whether a designated pause annotation (e.g., `crossplane.io/paused`) exists on the managed resource. If this annotation exists, all create/update/delete/generic events are filtered before they are delivered to the event handler. If not, all events are delivered by this predicate.

The issues with this approach are:
- If the managed resource chooses not to install the predicate, then it would not be adhering to XRM (as it would not be honoring the pause annotation). I think this is not a big concern, although I also believe if possible, we should be enforcing managed resource's to adhere to XRM. I believe this is not a big concern because a managed reconciler's own implementation may also not honor the external-name annotation, or may not do late initialization, etc., i.e., there are already existing aspects of XRM that we already don't enforce. But it looks like we can already enforce adherence to the pause annotation in the managed reconciler so I mentioned this as an issue here.
- More importantly, the managed reconciler already periodically enqueues reconciliation requests in the worker queue by setting a non-zero valued `reconcile.Result.Requeue` or `reconcile.Result.RequeueAfter`. And a controller-runtime predicate cannot interfere with this explicit queueing as it's not in the path between the event source and the event handler.

So, even if we make use of predicates, we would also need to disable requeueing in the managed reconciler. But then, we could just implement pausing explicitly at the managed reconciler level so that:
- If a managed resource is depending on a crossplane-runtime version with pause-annotation support, we would be enforcing adherence to pause-annotation for the managed resource. No extra plumbing would be necessary on the managed resource side.
- The implementation would just stay at the managed reconciler level and not be divided between controller-runtime & crossplane-runtime levels.

For implementing this at the managed reconciler level, we could employ an encapsulating reconciler implementation that would check the existence of the `crossplane.io/paused` annotation and return early if it exists, or calls the actual downstream reconciler, (in our case, this would probably be the rate-limiter reconciler), the annotation is not there. 

Another option could just be extending the logic in the current managed reconciler and implement the early return logic there.

**Comment by muvaf** [comment]
Thanks for the write up @ulucinar ! I wasn't aware of the fact that explicit re-queue cannot be blocked by a predicate. I agree that having a condition in the beginning of the reconciler would probably be the best way to go here for the reasons you listed.

**Comment by Feggah** [comment]
Maybe I am to late to the discussion @muvaf @ulucinar, but why the pause/resume reconciliation specification were designed to be an annotation?

Shouldn't it be part of the `spec` object? Something like `spec.reconcilePolicy: Enabled|Paused`.

Why is the "reconcile policy" set in the `metadata` and the deletion policy in the spec? IMO, the design for these features should have the same treatment.

> I will reopen the issue just for visibility

**Comment by muvaf** [comment]
@Feggah I think it's mostly because the main motivation is to enable smoother operation for migration which is a very infrequent operation. If the use cases grow and we see it needed frequently or with different combinations, we could definitely consider making it part of the spec, similar to `deletionPolicy` as you described. Separately, I think it's also a practice in upstream Kubernetes as well to start with an annotation and assess whether the field should go to spec.
